### PR TITLE
neomutt module

### PR DESCRIPTION
**Pull request by MasseR**
Add a new module for neomutt.

I'm still in the process of testing if all my use cases are handled by this module, but in the meantime if you could have a look over if there is something obvious.

I tried incorporating rycees guidance on the previous pull requests as well.

### CODE REVIEW

**Comment by MasseR** [comment]
This might actually be sort of overlapping with #225 but is a bit more self-contained.

**Comment by MasseR** [comment]
I have converted all of my mutt configurations to use this module. Seems to work at least for me.

File: modules/programs/neomutt.nix [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        type = types.string;
+        default = "";
+        description = "IMAP username.";
+      };
+
+      password = mkOption {
+        type = types.nullOr types.string;
+        default = null;
+        example = "`${pkgs.gnupg}/bin/gpg --decrypt passfile.gpg`";
+        description = "IMAP password. Because these configuration files are kept in the nix store unencrypted, I recommend using some kind of a password command.";
```
Comment (rycee) [inline modules/programs/neomutt.nix]: Please avoid such long lines. I much prefer something along the lines of
```
        description = ''
          IMAP password. Because these configuration files are kept in
          the nix store unencrypted, I recommend using some kind of a
          password command.
        '';
```
Also, instead of "I recommend" maybe "it is recommended".

File: modules/programs/neomutt.nix [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        type = types.string;
+        default = "";
+        description = "SMTP username.";
+      };
+
+      password = mkOption {
+        type = types.nullOr types.string;
+        default = null;
+        example = "`${pkgs.gnupg}/bin/gpg --decrypt passfile.gpg`";
+        description = "SMTP password. Because these configuration files are kept in the nix store unencrypted, I recommend using some kind of a password command.";
```
Comment (rycee) [inline modules/programs/neomutt.nix]: Also too long.

File: modules/programs/neomutt.nix [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+      width = mkOption {
+        type = types.int;
+        default = 22;
+        description = "Width of the sidebar";
+      };
+
+      shortPath = mkOption {
+        type = types.bool;
+        default = true;
+        description = "By default sidebar shows the full path of the mailbox, but with this enabled only the relative name is shown.";
```
Comment (rycee) [inline modules/programs/neomutt.nix]: Long line.

File: modules/programs/neomutt.nix [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+        type = types.bool;
+        default = false;
+        description = "If set, neomutt will attempt to use the IDLE extension.";
+      };
+
+      url = mkOption {
+        type = types.string;
+        default = "";
+        example = "mail.gmail.com";
+        description = "IMAP address without the `imaps://`.";
```
Comment (rycee) [inline modules/programs/neomutt.nix]: Instead of
```
`imaps://`
```
maybe
```xml
<literal>imaps://</literal>
```
since the description is interpreted as docbook.

File: modules/programs/neomutt.nix [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        ];
+        default = "threads";
+        description = "Sorting method on messages.";
+      };
+
+      mailboxes = mkOption {
+        type = types.listOf types.string;
+        default = [];
+        example = ["+github" "+Lists/nix" "+Lists/haskell-cafe"];
+        description = "A list of mailboxes. If you prepend the path with `+` the path will be considered to be after the current folder.";
```
Comment (rycee) [inline modules/programs/neomutt.nix]: `<literal>+</literal>`

File: modules/programs/neomutt.nix [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+        ${optionalString cfg.gpg gpgSection}
+
+        ${optionalString (cfg.checkStatsInterval != null) mailCheckSection}
+
+        set implicit_autoview = yes
+
+        alternative_order text/enriched text/plain text
+
+        set header_cache = "~/.mutt/cache/"
```
Comment (rycee) [inline modules/programs/neomutt.nix]: I'd suggest
```
set header_cache = ${config.xdg.cacheHome}/neomutt/header
set message_cachedir = ${config.xdg.cacheHome}/neomutt/message
```

**Comment by rycee** [comment]
Wow, impressive! I made a few comments but will need some more time to have a thorough look.

Ping for mail guru @teto as well 🙂

**Comment by MasseR** [comment]
I quickly edited the docs and file locations, give me some time to test these changes tomorrow and I'll also squash the commits back.

**Comment by MasseR** [comment]
And as for the #225 I think this might work well as an underlying primitive for the all-encompassing mail configuration.

**Comment by MasseR** [comment]
Still works :+1:

**Comment by teto** [comment]
@MasseR @rycee How do you want to proceed about this ? The goal of the mail PR is exaclty not to have to define MXA-specific (MUA/MRA/MTA) specifics. If we merge this now than we might have to break the API later.
Maybe I can strip down my PR of what is controversial if @MasseR is ok to rework the neomutt configuration to adopt the generic framework ? As the API might change with additionnal feedback/MUA/MRA/MTAs maybe we can do this in a "mail" branch and once it's a bit more tested merge it back to master.

**Comment by MasseR** [comment]
At least I find it extremely valuable to have the ability to do MUA/MRA/MTA specifics. Even if I didn't want to tailor my getmail/fetchmail specifically, I still want to configure neomutt to my liking.

I see value in having best of both worlds. General stuff for the well general stuff, but also program specific stuff for the nitty gritty details.

For the record, this is pretty much how I generalize my configuration between machines right now:

- I have a non-committed `credentials.nix` that contains the specifics for a given environment
- I have a committed `mail.nix` that uses the `credentials.nix` for server specific stuff, but sets up mutt and mbsync with the shared config as well as nitty gritty details.

~~~
# credentials.nix
{
  name = "John Doe";
  smtp_user = "john.doe";
  smtp_password = "`gpg --decrypt foo.gpg`"
  smtp_url = "smtp.gmail.com";
  imap_user = "john.doe";
  imap_password = "`gpg --decrypt bar.gpg`";
  imap_url = "imap.gmail.com";
  mailboxes = [ "INBOX" "nix" "haskell" ];
}
~~~

~~~
# mail.nix
{config,pkgs,...}:
let credentials = import ./credentials.nix;

in

{
  programs.mutt = {
    enable = true;
    imap = {
      enable = true;
      url = credentials.imap_url;
      user = credentials.imap_user;
      ...
    };
    smtp = {
      ...
    };
    sidebar = {
      ...
    };
    binds = [ ... ];
    macros = [ ... ];
    theme = ./theme;
    mailboxes = credentials.mailboxes;
  };
  programs.mbsync = {
    # pending for conversion to the newish home-manager module, but would use the imap username, password, url and mailboxes
  };
  programs.getmail = {
    # On one of my other machines I don't want to use mbsync, I use getmail instead
    retriever = {
     user = credentials.imap_user;
     ...
    };
  };
}
~~~

**Comment by teto** [comment]
> I see value in having best of both worlds. General stuff for the well general stuff, but also program specific stuff for the nitty gritty details.

That's the idea. The main goal is to factorize the account invariants such as login/password/gpg key/signatures details so that can switch between different programs without grind.
I can't bear having to repeat server details across neomutt/vdirsyncer/offlineimap/mbsync/astroid/alot. So in your example this would only concern this part that could be moved to an `accounts` attrset

```
    imap = {
      url = credentials.imap_url;
      user = credentials.imap_user;
      ...
    };
    smtp = {
      ...
    };
```

**Comment by MasseR** [comment]
Yeah. IMO either leave the responsibility of using the accounts attrset to the caller, or have an **overrideable** value in the neomutt configuration.

Few things to consider

- I prefer not having passwords in the nix store -> use a password command if only supported
- Some applications have a different style for using commands
  - mutt: quoted backticks
  - mbsync: a field with a plus prefix and quotes `PassCmd +"gpg --decrypt foo.gpg"`
  - getmail: a tuple of cmd and arguments `password_command = ("gpg", "--decrypt", "foo.gpg")`

With the exception of passwords, using the values from your accounts attrsets in these program/service primitives would be easy to add, even later on.

So in short, my recommendation is to have these few programs as primitives that use your account attrset once it's eventually merged. This change would be done either by you or me.

**Comment by rycee** [comment]
Howdy. Was thinking about this over the weekend and put together #295 as a sort of proof of concept. It's not very fleshed out but I'm hoping the general idea is flexible enough to allow general mail account configuration while still allowing detailed per-program configuration without making the mail module itself too cluttered.

**Comment by rycee** [comment]
@MasseR @teto I have merged an initial email infrastructure setup to master now.

I included two modules, notmuch and mbsync, as example usage. I've also reworked the neomutt module of this PR to use the new email module: #300

**Comment by teto** [comment]
cool ! Thanks for the work on the account module. I've started looking into it and adapted the msmtp module. I am now looking at upgrading the alot plugin and how to deal with multiple accounts. My previous PR did current . When looking at the new notmuch/neomutt package, I notice that every account seems recorded in the same file ?  I liked having one configuration per account ( also because some software won't support several accounts per config file). For neomutt can you switch between accounts defined in the same config ?

Also the pattern to enable per-application is to add booleans like `myMailAccount.application.enable = true;`. I wonder if it wouldn't be better from a discovery standpoint to have mail applications accept instead an "accounts" option (that defaults to all accounts). This way it gets picked up by the documentation generator (unlike now).

**Comment by teto** [comment]
Seems like it is in fact already discoverable, my bad:
```
+    accounts.email.accounts = mkOption {
+      options = [
+        {
+          neomutt = {
+            enable = mkOption {
```

Now I wonder about to generate the commands called by the mua for the mra/mta.
should we add `neomutt.mta = mkOption {   type = types.enum [ "msmtp" ]; }` ?

**Comment by rycee** [comment]
@teto Thanks for the feedback. Yeah, I considered having something like, e.g., `programs.notmuch.accounts` but eventually opted against it since I found it more natural to have everything center around the account and the account configuration rather than the programs. I believe it also allows the programs to more easily make use of the "flavor" option.

About the single vs. multiple configurations I had a look at different ways to do it and checked the documentation of mbsync, notmuch, and neomutt. It seems to me that the idiomatic way to set up multiple accounts in these programs is to have a single configuration. My hypothesis here is that one should easily be able to work with multiple accounts from the same program session whenever possible. For example, to search across all accounts in a single notmuch query or to switch between accounts in neomutt without having to restart it. Note, the PR I set up for neomutt has basic support for multiple accounts.

From what I can tell msmtp also supports multiple accounts in one configuration and will choose based on the From field. But yeah, I guess plenty of programs still only support one account at a time and then I think having multiple configs is perfectly fine.

About neomutt and configuration of mta I can think of two reasonable designs,

1. Automatically using `msmtp` if `accounts.email.accounts.<name>.msmtp.enable = true`, otherwise use the builtin SMTP support, or

2. Have an option `accounts.email.accounts.<name>.neomutt.mta` of type `enum [ "builtin" "msmtp" ]`, defaulting to "builtin".

**Comment by rycee** [comment]
@teto My understanding of Notmuch is that everything is done with tags 🙂 So in your case I think something like
```nix
{
programs.notmuch.new.tags = [ "new" ];
programs.notmuch.hooks.postNew = ''
  notmuch tag +personal -- tag:new and \( from:teto@personal.com or to:teto@personal.com \)
  notmuch tag +professional -- tag:new and \( from:teto@professional.com or to:teto@professional.com \)
  notmuch tag +inbox +unread -new -- tag:new
'';
}
```
would be the idiomatic way to distinguish between private and professional email. If you prefer wrapper scripts then they could inject the tag into the search query, something like
```bash
#! /usr/bin/env bash
cmd="$1"
shift
notmuch $cmd tag:personal and \("$@"\)
```
perhaps.

**Comment by teto** [comment]
I thought about that and even though I might end up using a single db, here are some points:
- my  impression is that the notmuch config can contain only one identity (realName) but I don't want to use my real name for personal mail
- my mua `alot` is not capable of switching between different files
- Keeping pro/perso in separate databases is a good way to  prevent mistakes upon contact completion, uninteresting results in research etc. Enforcing the separation would force me to add labels for every mail/search...
I think keeping separate databases should be possible, maybe via an option.

**Comment by rycee** [comment]
Yeah, notmuch is a bit limited in that it only handles one name. On the other hand, I had a look at the notmuch source code and the name doesn't seem to be used very much. The only serious use I saw was in the `notmuch reply` functionality, which uses it to construct the `From` field. I don't know how frequently used the reply command is, do you use it?

About "my mua alot is not capable of switching between different files" I'm not certain what you mean. FWIW I had a look at the alot manual and it does seem to have support for [accounts](https://alot.readthedocs.io/en/latest/configuration/accounts.html), at least and [per-account address books](https://alot.readthedocs.io/en/latest/configuration/contacts_completion.html).

In any case, I have no issue against notmuch being configurable to optionally support separate configuration files per account. But I guess the design has to be quite careful. I don't want the modules to gain much in complexity. For example, if alot is added, how should it be set up to cope with the two cases? Same if notmuch support is added to the neomutt module or a future Emacs notmuch configuration…

**Comment by teto** [comment]
closing in favor of https://github.com/rycee/home-manager/pull/300 (to reduce confusion).
